calendra/core.py

```diff
@@ -506,12 +526,231 @@ def get_working_days_delta(self, start, end, include_start=False):
+            Default is [2000, 2030]
+
+        target_path: str
+            the name or path of the exported file. If this argument is missing,
+            the function will return the ical content.
+
+        """
+        first_year, last_year = self._get_ical_period(period)
+        if target_path:
+            # Generate filename path before calculate the holidays
+            target_path = self._get_ical_target_path(target_path)
+
+        # fetch holidays
+        holidays = []
+        for year in range(first_year, last_year + 1):
+            holidays.extend(self.holidays(year))
+
+        # initialize icalendar
+        ics = [
+            'BEGIN:VCALENDAR',
+            'VERSION:2.0',  # current RFC5545 version
+            f'PRODID:-//workalendar//ical {__version__}//EN'
+        ]
+        common_timestamp = datetime.utcnow().strftime('%Y%m%dT%H%M%SZ')
+        dtstamp = 'DTSTAMP;VALUE=DATE-TIME:%s' % common_timestamp
+
+        # add an event for each holiday
+        for holiday in holidays:
+            date_ = self.get_observed_date(holiday)
+            ics.extend([
+                'BEGIN:VEVENT',
+                'SUMMARY:%s' % holiday.name,
+                'DTSTART;VALUE=DATE:%s' % date_.strftime('%Y%m%d'),
+                dtstamp,
+                f'UID:{date_}{holiday.name}@peopledoc.github.io/workalendar',
+                'END:VEVENT',
+            ])
+
+        # add footer
+        ics.append('END:VCALENDAR\n')  # last line with a trailing \n
+
+        # Transform this list into text lines
+        ics = "\n".join(ics)
+
+        if target_path:
+            # save iCal file
+            with open(target_path, 'w+') as export_file:
+                export_file.write(ics)
+            return
+
+        return ics
+
+
+class Calendar(CoreCalendar):
+    """
+    The cornerstone of Earth calendars.
+
+    Take care of the New Years Day, which is almost a worldwide holiday.
+    """
+    include_new_years_day = True
+    shift_new_years_day = False
+    include_labour_day = False
+    labour_day_label = "Labour Day"
+
+    def __init__(self, **kwargs):
+        super().__init__()
+
+    def get_fixed_holidays(self, year):
+        days = super().get_fixed_holidays(year)
+        if self.include_new_years_day:
+            days.insert(
+                0, (date(year, 1, 1), "New year")
+            )
+
+        if self.include_labour_day:
+            days.append(
+                (date(year, 5, 1), self.labour_day_label)
+            )
+        return days
+
+    def get_variable_days(self, year):
+        days = super().get_variable_days(year)
+        new_year = date(year, 1, 1)
+        if self.include_new_years_day and self.shift_new_years_day:
+            if new_year.weekday() in self.get_weekend_days():
+                days.append((
+                    self.find_following_working_day(new_year),
+                    "New Year shift"))
+        return days
+
+
+class SeriesShiftMixin:
+    """
```

Comment:
@jaraco I wanted to draw your attention to this new Mixin. Hopefully, I explained it's purpose but if I was not clear enough, please ask.

Reply:
Thanks for pointing this out. Is this change necessary to support the newer workalendar? It would be nice to consider this new feature (and divergence from workalendar) independently from the sync with workalendar. If it's not strictly necessary to get a synced version to work, then I'd recommend considering this separately from the rest of the sync. If it is necessary, would it be possible to contribute this functionality to calendra master prior to syncing with workalendar?

The current calendra already deals with some small series (Thanksgiving in US, Christmas+Boxing day in UK). I see Thanksgiving doesn't bother with shifts because it always falls on a consistent day of the week. But [here is how Boxing Day indicates that it falls after Christmas](https://github.com/jaraco/calendra/blob/fa03ff090d39e85ed2099571e14b1e8f3ec17944/calendra/core.py#L653). In this way, Boxing day knows to honor both Christmas and its shift relative to any given calendar.

Ideally, a "Series" would be handled similarly - where each day is a Holiday, and each holiday is aware of its dependencies. Even better might be to reflect the reality of the situation and let a single Holiday be observed on multiple days (and to allow that holiday to set its rules for when it's observed, relative to the configuration of the calendar).

In other words, this hook point feels like not the best place to model the behavior. I'd hope that there could be a cleaner way than keeping track of series by label and copy/pasting the observance shift from each.

I'll have to think about this some more, but please let me know what you think about my questions above.

Reply:
Although I've expressed some reluctance here, I suspect this approach may be suitable at least for an initial implementation. Please answer the questions if you can, but don't feel discouraged. Let's focus on getting the tests passing and iterate from there.

Reply:
Good questions, let me start from where I found the issue...

When trying to sync up to the upstream v11.0.0 release, I was getting test failures for China and some Islamic countries (Turkey, Malaysia adn Qatar). While poking around these Islamic failures, I realised that the two Eid holiday "series" were not being shifted correctly: only the day in the series that actually fell on a Sunday was being shifted, and that shift then caused it to be a duplicate of the next day in the Series.

This then fell foul of the Workalendar support for ical export (introduced circa Aug 2020). This includes tests which assert, amongst other things, on the uniqueness number of `uid_lines`...which the duplicates now broke.

I realised at this point that I could not see any logic in either Workalendar or Calendra to deal with this - except that I had already seen the custom logic that handled Christmas day and Boxing day (recall issue #15 is why I was here in the first place).

Note: since we have your observance day logic as distinct from what upstream does  (which is why I choose to use Calendra over Workalendar), there are two different codebases with the same problem! I immediately realised that I had little prospect of a decent/viable fix for upstream, but your logic provides easy detection AND a basis for a somewhat tacky solution. The current implementation is a direct consequence of this train of thought.

(Once I implemented it for Malaysia, I realised the same fix was needed for the other Islamic countries, and shortly later for the various series holidays in China).

The shortcomings of the solution are:

1. No obvious backport upstream.
2. As you point out, the lack of a code structure corresponding to the problem in hand.  Conceptually at least, this is broken since it cannot handle the obvious possibility that the series itself may have a knock on effect on a later holiday.
3. No immediate backport to the Christmas/Boxing day series. This is a direct consequence of the conceptual problem from 2.
4. I especially don't like the need to tailor the copy-pasting of the observance for each year in case the shift point moves, and all the duplication of Holiday. (Part of that at least could be optimised but...).

I felt obliged to ignore 1, and the solution for 3 would have been intrusive. As for 2, I considered subclass of Holiday to wrap the series, but could not see a comfortable fit for the observed_date() API:

- A covariant return like `Union[date, List[date]]` seemed horrid
- A new observed_dates() API seemed worse.

So that's why I ended up with what you see.

Finally, with hindsight, it would definitely have been better as a separate commit ahead of the sync. But I didn't think about that at the time, and not having the fix meant that v11.0.0's tests would not pass. Moving it now would be tedious, but may be the right thing to do (not least because it would make filing a bug upstream easier). Your call...

Reply:
Thanks for the response Shaheed. Your clarity and diligence and eloquence in helping me understand the situation is exceptional.